server/halomod_app/utils.py

```diff
@@ -0,0 +1,362 @@
+                        getattr(o, x)[mask],
+                        y[mask],
+                        color=f"C{i % 7}",
+                        linestyle=lines[(i // 7) % 4],
+                        label=l,
+                    )
+            except Exception as e:
+                logger.exception(f"Error encountered getting {q} for model called {l}.")
+                errors[l] = e
+        else:
+            if i == 0:
+                comp_obj = o
+                continue
+
+            try:
+                ynum = getattr(o, q)
+            except Exception as e:
+                logger.exception(f"Error encountered getting {q} for model called {l}.")
+                errors[l] = e
+
+            yden = getattr(comp_obj, q)
+            mask = yden > 0
+
+            if ynum is not None and yden is not None:
+                y = ynum[mask] / yden[mask]
+                ys[l] = y
+                ax.plot(
+                    getattr(o, x)[mask],
+                    y,
+                    color=f"C{(i+1) % 7}",
+                    linestyle=lines[((i + 1) // 7) % 4],
+                    label=l,
+                )
+
+    try:
+        # Shrink current axis by 30%
+        ax.set_xscale("log")
+
+        ax.set_yscale(d["yscale"], base=d.get("basey", 10))
+        if d["yscale"] == "log" and d.get("basey", 10) == 2:
+            ax.yaxis.set_major_formatter(tick.ScalarFormatter())
+
+        box = ax.get_position()
+        ax.set_position([box.x0, box.y0, box.width * 0.6, box.height])
+
+        # Put a legend to the right of the current axis
+        ax.legend(loc="center left", bbox_to_anchor=(1, 0.5), fontsize=15)
+
+        buf = io.BytesIO()
+
+        if plot_format == "pdf":
+            FigureCanvasPdf(fig).print_pdf(buf)
+        elif plot_format == "png":
+            FigureCanvasAgg(fig).print_png(buf)
+        elif plot_format == "svg":
+            FigureCanvasSVG(fig).print_svg(buf)
+        else:
+            raise ValueError("plot_format should be png, pdf or svg!")
+    except Exception:
+        logger.info(f"y-axis data: {ys}")
+        logger.exception("Something went wrong in creating the image itself")
+        raise
+
+    return buf, errors
+
+
+MLABEL = r"Mass $(M_{\odot}h^{-1})$"
+KLABEL = r"Wavenumber, $k$ [$h$/Mpc]"
+RLABEL = r"Scale, $r$ [Mpc/$h$]"
+KHMLABEL = r"Fourier Scale, $k$ [$h$/Mpc]"
+
+XLABELS = {"m": MLABEL, "k": KLABEL, "r": RLABEL, "k_hm": KHMLABEL}
+
+KEYMAP = {
+    "dndm": {
+        "xlab": MLABEL,
+        "ylab": r"Mass Function $\left( \frac{dn}{dM} \right) h^4 Mpc^{-3}M_\odot^{-1}$",
+        "yscale": "log",
+    },
+    "dndlnm": {
+        "xlab": MLABEL,
+        "ylab": r"Mass Function $\left( \frac{dn}{d\ln M} \right) h^3 Mpc^{-3}$",
+        "yscale": "log",
+    },
+    "dndlog10m": {
+        "xlab": MLABEL,
+        "ylab": r"Mass Function $\left( \frac{dn}{d\log_{10}M} \right) h^3 Mpc^{-3}$",
+        "yscale": "log",
+    },
+    "fsigma": {
+        "xlab": MLABEL,
```

Comment:
all of this should be somewhere in the client-side code, right?

Reply:
It is pulled in to the client side code from the server. In a way it is there. This would be good to discuss at the meeting tomorrow though.